### DISCUSSION TITLE
HTTP2 streams

### DISCUSSION THREAD

**Discussion by Modcat** [feathersjs/feathers#2362] [Ideas]
Would it possibly for you guys to start using HTTP2 streams instead of sockets, all web servers support HTTP2 and we don't have to downgrade in our configs, plus it'd be easier and more up to date on server and client side without the need for all this socket.io stuff!

**Comment by J3m5** [comment]
You mean server sent events ?
Why do you have to downgrade in your configs ? What does it means ?

Technically, server sent events could be used to replace websockets, we already talked about it several times,
this would need to be implemented.

**Reply by Modcat** [reply]
To run websockets we have to downgrade to HTTP 1.1 FYI from HTTP2 which makes certain things slower when working with websockets than out of the box. Also HTTP 3 is now supports and does offer an alternative to websockets.

**Reply by J3m5** [reply]
Ok, I see.
But in the mean time, HTTP and HTTP2 could be used side by side, no ?
At least it's not a problem for [caddy for example](https://caddy.community/t/disabling-http-2-for-websocket/2437/3) and it shouldn't be for feathers too.
Put clearly, IMO, you shouldn't have to downgrade all the connections to HTTP1 in order to be able to use websockets as they are distinct transports and nothing prevent a browser to make HTTP1 websocket requests and http2 requests at the same time.

**Reply by daffl** [reply]
With websockets, HTTP is only used for the one handshake request to establish the connection. Once the connection is established everything goes through the socket. Other clients should still be able to make HTTP 2 REST requests at the same time.

Server sent events would be a nice feature but it seems like there never was much demand for it. One of the things I learned over the years is to be considerate about adding new integrations and transports because it is a lot of work that often doesn't get adopted (the vast majority just uses whatever is recommended in the CLI).

**Comment by Modcat** [comment]
Also I would like to add something here @daffl @J3m5 there is a really cool API already available for this called gRPC https://grpc.io/ which has support for nodeJS out of the box!
